#644, Recursive filtering

K1rakishou opened the pull request:
Closes #644.

Added two new menu items - "Hide" and "Hide whole chain". "Hide" is for hiding separate posts/threads whereas "Hide whole chain" is for hiding posts with all replies to those posts. OP post cannot be hidden from a thread (only from the catalog and in this case the whole thread will be hidden).

Adamantcheese commented:
This has LOTS of issues with crashing, especially when scrolling fast through a thread or reaching the last post in a thread (in which case it crashes immediately when a thread has only a few posts and can't scroll yet).

K1rakishou commented:
@Adamantcheese Alright, I will look into it today.

K1rakishou commented:
When I tested this I had crashes in this place so I kinda closed it up with that check. Maybe it is related.
https://github.com/Floens/Clover/pull/654/files#diff-c43fe9f6e874c2b42ffaf3822bf843edR107

Adamantcheese commented:
It will also crash here with the opposite error as the one you fixed. https://github.com/K1rakishou/Clover/blob/(%23644)-recursive-filtering/Clover/app/src/main/java/org/floens/chan/ui/adapter/PostAdapter.java#L129
i.e.
```if (!(holder instanceof StatusViewHolder)) {
                    // to avoid "PostViewHolder cannot be cast to StatusViewHolder" exception
                    return;
}
will also "solve" a problem, but the crashes will continue. It's likely that these two checks aren't needed and there's something else wrong with your code.

K1rakishou commented:
@Adamantcheese I think I figured out what is the problem.

https://github.com/K1rakishou/Clover/blob/(%23644)-recursive-filtering/Clover/app/src/main/java/org/floens/chan/ui/adapter/PostAdapter.java#L198

Here `displayList` is being accessed on the background thread and then there is the `notifyDataSetChanged` which will trigger `onBindViewHolder` to be called and it will access the `displayList` on the main thread. So, to fix this `displayList` should be synchronized. We can probably completely remove `sourceList` since it's not being used anywhere besides this place. The loop before `notifyDataSetChanged` call should be synchronized as well. I will push the fixes a little bit later today, have some stuff to do. Or you can just apply this patch.

[#654_fix.zip](https://github.com/Floens/Clover/files/3077187/654_fix.zip)

Adamantcheese commented:
It's about 2AM for me, so whenever you get to it, I'll test it and make sure it's good before putting it in. 

Although why does this need to be done in a background thread? The current version works fine with it running on the main thread and the synchronization is going to be a pain in the ass to make work correctly (as I figured out when trying to resolve this).

If I remember right, most of the crashes were due to IndexOutOfBounds exceptions. It also looks like [another reason it might fail is here](https://github.com/K1rakishou/Clover/blob/(%23644)-recursive-filtering/Clover/app/src/main/java/org/floens/chan/ui/adapter/PostAdapter.java#L201) where the background thread is resetting the indicator position before calculating it, instead of using a thread-safe local variable to calculate it before setting it. That -1 being set for (relatively) quite a while, considering the holder stuff seems to wait in nanoseconds before failing, would explain those exceptions. If you do choose to continue using a background thread, you should probably double check that because it might cause crashes still.

K1rakishou commented:
@Adamantcheese 

>Although why does this need to be done in a background thread?
Before this task you could only hide threads. Hidden threads were stored in the database as well as in a cache variable for fast look-ups (https://github.com/Floens/Clover/pull/654/files#diff-d4c6b656f34af7234a6390b559ec3688L21). That's why it could work safely on the main thread. But it could only hold 250 threads at a time. After I've implemented the ability to hide posts as well as threads the number of 250 hidden posts became kinda small. So I've removed those caches and instead filtering is now reading hidden posts directly from the database. And now it can't run on the main thread since it will just crush (or just hung for a while if there is an ability to disable throwing an exception when accessing database from the main thread in the orm lite).

I think you are right that `lastSeenIndicatorPosition` may cause different crashes so what I think to do is to completely extract this code from the adapter (it shouldn't be the in the first place) to a different place and then just assigning the filtering result into the adapter on the main thread. I'm gonna work on it later today.

K1rakishou commented:
Alright. This should fix it.

myuli commented:
Hey @K1rakishou , some threads hide some don't.

Also, could you make it that instead of removing the posts (inside a thread) they would just hide?
Dashchan does it like you hide them but they stay in place and if you click them they appear again.

![Dashchan-Hide](https://user-images.githubusercontent.com/33556300/56116593-da2b1600-5f2b-11e9-8bac-3ad8cbaa6a5d.png)

Astridchan commented:
Wouldn't it make sense to add options for "remove whole chain" and "hide whole chain" in the filter settings? As currently the recursive filtering feature is only available when manually hiding posts.

![Screenshot_2019-04-15-12-39-42-755_org floens chan](https://user-images.githubusercontent.com/40862101/56122721-b394cb80-5f7b-11e9-8481-1f2b802a249f.png)

Also, Muyli above has a point regarding hiding vs. removing posts. Currently the options in filter settings are "hide" and "remove", where "hide" will gray out the post and "remove" will completely remove it. But when you click the three button menu on a post, the "hide" option doesn't "hide" the post, it removes it. So the same term "hide" is being used for two different actions, which I think can lead to confusion. My suggestion would be to add both "remove" and "hide" in the post menu, and to keep the menu from getting cluttered, instead of having a "hide whole chain" and "remove whole chain" option separately, you would get and Y/N prompt "hide/remove whole chain?" when pressing to hide or remove a post with replies.

K1rakishou commented:
@myuli @Astridchan 
So, to sum it up. There have to be two menu items - "Hide" which will grey out the post and "Remove" which will completely remove the post like it is implemented right now. Upon clicking a post if it has replies to it a dialog windows should pop up suggesting the ability to hide/remove the whole chainю Is this correct?

>Wouldn't it make sense to add options for "remove whole chain" and "hide whole chain" in the filter settings? As currently the recursive filtering feature is only available when manually hiding posts.
Wouldn't people start complaining that posts are disappearing without any reason because they forgot this option is turned on?

Astridchan commented:
@K1rakishou 
> So, to sum it up. There have to be two menu items - "Hide" which will grey out a post and "Remove" which will completely remove a post like it is implemented right now. Upon clicking a post if it has replies to it a dialog window should pop up suggesting the ability to hide/remove the whole chain. Clicking on hidden (not removed) post should bring it back up. Is this correct?

Exactly.

> Wouldn't people start complaining that posts are disappearing without any reason because they forgot this option is turned on?

Well that's on them, isn't it?

I'll try to explain why I think this feature would be useful, and why I would want to see it added. Basically, in a few of the generals I frequent, there are these copy-pasta shitposts that get posted in every new thread for (You)s. I've got those posts filtered so that they're automatically removed, but the replies remain, so I still end up seeing the shitposts when I follow back the reply chains. A "remove whole chain" would make it so both the shitpost and the replies to it get removed.

K1rakishou commented:
The last commit is Work In Progress but you can already test it if you would like (I would really appreciate it).

It "should" work. 
For now all posts that have replies to a post that was matched by a filter will be hidden (greyed out). I'm gonna need to store the filter id along with the posts to figure out what to do with such a post (hide/remove or highlight) when filtering them.

K1rakishou commented:
Alright, looks good.
So, now there are two menu options - hide and remove. 
- If a post has replies to it then additional dialog will be shown with suggestion to remove/hide the whole chain. 
- If you choose to remove a single post and then someone replies to this post the reply will NOT be hidden (idk should I implement it?).
- If a post has replies and you choose to hide/remove the whole chain then all new replies to any of the post from the chain will be hidden/removed.
- Filters now have an additional parameter "Apply to replies". 
![234234](https://user-images.githubusercontent.com/10332458/56472109-3d083a00-6463-11e9-8c32-0272743b96e4.png)
- If anyone replies to a post that is hidden by a filter with this flag it will be hidden/removed/highlighted as well. The action (hide/remove/highlight) is inherited from the parent post. So if you have a filter that highlights posts and someone replies to a post that was hidden by that filter that post will be highlighted as well.
- Hide/remove menu options are disabled for posts that has been affected by a filter.

I changed the DB again and have no idea how to handle exporting/importing from #678 so that @Adamantcheese build works correct because I will need to merge this task with #678 but then this task will grow up in size. We really need @Floens to start merging PRs.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/site/parser/PostParseCallable.java:
```diff
@@ -77,15 +77,18 @@ private void processPostFilter(Post.Builder post) {
             if (filterEngine.matches(filter, post)) {
                 FilterEngine.FilterAction action = FilterEngine.FilterAction.forId(filter.action);
+
+                // TODO: store filter id in the post so we can figure out what exactly we want to do
```
should remove this

Adamantcheese commented:
I'll take a look at what changed in the db and do what I need to for import/export. I have an exported file for the last version that I can test with.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/database/DatabaseHideManager.java:
```diff
@@ -1,118 +1,399 @@
     }
 
-    public Callable<Void> clearAllThreadHides() {
+    public Callable<Void> removeThreadHide(PostHide hide) {
```
Probably should do the same as in removePostsHide

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/database/DatabaseHideManager.java:
```diff
@@ -1,118 +1,399 @@
+    /**
+     * Rebuilds a child post with custom filter parameters
+     * */
+    private Post rebuildPostWithInheritedFilter(
```
rename to rebuildPostWithCustomFilter

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/model/Post.java:
```diff
@@ -244,6 +250,34 @@ public PostImage image() {
+            return false;
+        }
+
+        if (this.getClass() != other.getClass()) {
```
Add this == other check

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/model/Post.java:
```diff
@@ -244,6 +250,34 @@ public PostImage image() {
+
+    @Override
+    public int hashCode() {
+        return 31 * no;
```
Add board and site as well

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/model/Post.java:
```diff
@@ -244,6 +250,34 @@ public PostImage image() {
+            return false;
+        }
+
+        return this.no == ((Post) other).no;
```
Add board and site as well

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/model/Post.java:
```diff
@@ -244,6 +250,34 @@ public PostImage image() {
+
+    @Override
+    public String toString() {
+        return "[no = " + no + ", comment = " + comment + "]";
```
Add board and site as well

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/utils/PostUtils.java:
```diff
@@ -0,0 +1,135 @@
+    private PostUtils() {
+    }
+
+    public static Post findPostById(int id, ChanThread thread) {
```
Maybe pass a list of posts here instead of ChanThread so we can get rid of ChanThread dependency here?

K1rakishou replied:
Nah, it's more convenient this way.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/database/DatabaseHelper.java:
```diff
@@ -256,6 +259,36 @@ public void onUpgrade(SQLiteDatabase database, ConnectionSource connectionSource
+
+        if (oldVersion < 26) {
+            try {
+                postHideDao.executeRawNoArgs("ALTER TABLE threadhide RENAME TO " + POST_HIDE_TABLE_NAME + ";");
```
Woah, this is actually bad. Database migration queries should not have variables in them. If anyone renames at least one of them everything will break. Should replace them all. Idk how I didn't notice it before.

Adamantcheese commented:
@K1rakishou looks like from the last build of mine to the most recent one, no import/export changes needed to be made so the import/export version stayed the same. I've also already replaced all the database variable names in my build with their respective strings.

K1rakishou commented:
@Adamantcheese I see you have removed them from everywhere. I actually meant to just remove them from this class (https://github.com/Adamantcheese/Clover/blob/92a7e6d05778c6f304b4118be9c9de40d519a9e7/Clover/app/src/main/java/org/floens/chan/core/database/DatabaseHelper.java) because they should never change in migrations to avoid crashes, but keep them everywhere else to avoid mistakes when typing them by hand every time. And it is really helpful when you want to rename some column because you will only have to rename it in one place (where it is declared).

Adamantcheese commented:
I figured if everything was hardcoded it would be better and these names shouldn't really ever change.

K1rakishou commented:
@Adamantcheese Well normally yes, they should not change. Well I guess it's fine.

K1rakishou commented:
Regarding the importing/exporting I guess it is fine too because we have just merged the importing/exporting so there is just nothing to import from.

Adamantcheese commented:
The import/export was merged before this last set of changes you made, and the GSON class maker thing defaults the field to false if it didn't exist, so there's no upgrade needed.

Adamantcheese commented on Clover/app/src/main/java/org/floens/chan/core/model/orm/Filter.java:
```diff
@@ -49,6 +50,12 @@
     @DatabaseField(canBeNull = false)
     public int color;
 
+    @DatabaseField(canBeNull = false)
```
I meant to point to this line of code, sorry.

K1rakishou replied:
Done.

K1rakishou commented:
@Adamantcheese Damn, I guess I forgot to specify the name in the annotation. I wonder why it didn't crash for me when I was testing it.

myuli commented:
Yo @K1rakishou can you make it that hidden threads or replies unhide when you click on them?

K1rakishou commented:
@myuli Sure, but only for the posts that you hid by yourself and not by the filters. That's because of how filters work right now. I guess it should also be implemented for removed threads/posts.

K1rakishou commented:
Alright. You can now unhide threads/posts by clicking on them. 
There is also a new menu item that will show all posts that were removed in a thread.

![234234](https://user-images.githubusercontent.com/10332458/57086865-9ac04000-6d07-11e9-8b51-d5c8a62c2286.png)

myuli commented:
@K1rakishou Thanks, man. 
@Adamantcheese Can you add this?
